### ISSUE TITLE
Extracting route registration logic

### ISSUE THREAD

**Issue by Amit-PivotalLabs** [cloudfoundry/cf-mysql-release#85]
TL;DR: we want to change something you use, shouldn't affect you in the short run, but want to get your thoughts on the medium-term/long-term, since it will have some impact on this project.  We have [a story](https://www.pivotaltracker.com/story/show/110684012) in progress, blocked on this discussion.

We have built a general-purpose route registration job with a couple goals in mind:

- services that aren't primarily concerned with route registration logic shouldn't have to worry about that, and can simply colocate an "agent" if it's being deployed in a context where it wants to register a route, and the agent handles the logic for them.
- we have a common, reusable implementation so that there aren't 100 different ways to register routes, and 100 different things that need to be updated when route registration APIs change.

Currently it lives in cf-release, but the goal would be to move it into a routing-release.  Right now it's a simple Ruby process that talks to NATS, but we are switching it to use this Golang library that your project also uses: https://github.com/cloudfoundry-incubator/route-registrar.

We are working on that code to make it more general, and add some features to it (e.g. health checking, support for multiple routes to a single IP).  In the short run, this won't affect this project as long as you don't try to update your dependency to point to our new changes to that Golang library.  In the longer run, this will entail:

- cf-mysql-release deployments that live in the context of a gorouter, registering routes (presumably for the broker) will want to colocate this route_registrar job
- manifests will need to configure the route_registrar differently, for example [this](https://github.com/cloudfoundry/cf-mysql-release/blob/develop/jobs/cf-mysql-broker/spec#L64-L65), [this](https://github.com/cloudfoundry/cf-mysql-release/blob/develop/jobs/cf-mysql-broker/templates/registrar_settings.yml.erb#L7), and [this](https://github.com/cloudfoundry/cf-mysql-release/blob/develop/templates/cf-mysql-template.yml#L211) would go away, and the resulting configuration would look something more like [this](https://github.com/cloudfoundry/cf-release/blob/master/spec/fixtures/aws/cf-manifest.yml#L240-L241) and [this](https://github.com/cloudfoundry/cf-release/blob/master/spec/fixtures/aws/cf-manifest.yml#L216-L226)

Any thoughts or concerns?  We have [a story](https://www.pivotaltracker.com/story/show/110684012) in progress blocked on this discussion.

/cc @ljfranklin @menicosia @robdimsdale @zaksoup

**Comment by menicosia** [comment]
:+1:

I think the only quibble we might have is necessarily including all of cf-release in order to get at the single route-registrar job?

**Comment by Amit-PivotalLabs** [comment]
Right, which is why I mention that as the long-term picture, when it's
moved into the routing-release so you wouldn't pull in all of cf-release.
So in the short term, just don't do anything.  The only thing is, if in the
short term you think you'll need to change how you do route registration,
before the route_registrar job is available in routing-release, then you'll
have to adapt to some of the interface changes to the route-registrar
codebase.

Amit

**Comment by ljfranklin** [comment]
@Amit-PivotalLabs :+1: Sounds fine to me. We haven't had to make changes to route-registrar very often. If we do need to bump it for some reason, updating our release to conform to interface changes should be pretty quick.

Optimistically closing out this issue. Please re-open if there's more to discuss.

**Comment by Amit-PivotalLabs** [comment]
Great, don't think there's any more to discuss.

**Comment by Amit-PivotalLabs** [comment]
Actually @ljfranklin quick question, do you think any other projects depend on the route-registrar repo, who we should also be reaching out to?

**Comment by ljfranklin** [comment]
@Amit-PivotalLabs that is a good question. Looking at the closed issues in route-registrar, one of the teams in London might use it.

@benlaplanche @avade do you know if any of the London teams use our [route-registrar](https://github.com/cloudfoundry-incubator/route-registrar) submodule in their releases?

**Comment by benlaplanche** [comment]
Redis, RabbitMQ & Cassandra all use this, it sounds like no immediate impact for us. Unless we bump the dependency version in which case we'll update the manifest as well as per Lyles comments.

**Comment by ljfranklin** [comment]
Thanks Ben, closing this out.
